tools/driver/driver.cpp

```diff
@@ -134,6 +132,10 @@ int main(int argc_, const char **argv_) {
       llvm::cl::TokenizeGNUCommandLine,
       argv);
 
+  // Initialize the stack trace using the parsed argument vector with expanded
+  // response files.
+  PROGRAM_START(argv.size(), argv.data());
```

Comment:
It looks like this can be affected by subsequent operations on `argv`. Possible options are to push `PROGRAM_START` even further down in `main`, or to make a copy of `argv` at this location for the sole purpose of initializing the stack trace.

Reply:
Hm. Maybe we can do it the other way, and make sure that no one else changes argv after this point? The case where that does happen, `shouldRunAsSubcommand`, can be responsible for making the copy. What do you think?

Reply:
To make sure I understand your suggestion, the signature for `shouldRunAsSubcommand` would look like:
```
static bool shouldRunAsSubcommand(
  StringRef ExecName,
  SmallString<256> &SubcommandName,
  const SmallVectorImpl<const char *> &InputArgs,
  SmallVectorImpl<const char *> &OutputArgs) { ... }
```
`InputArgs` comes from what is currently `argv`. `OutputArgs` would initially point to `InputArgs`, but would be replaced with a brand new vector in the event of a modification. `OutputArgs` is what would be used in the rest of `main`. This wouldn't necessarily prevent someone from modifying `args`/`InputArgs`, however.

I'm not sure if there is a stronger way to ensure that the argument vector doesn't get modified, aside from:
```
const SmallVector<const char *, 256> trace_args(argv);
PROGRAM_START(trace_args.size(), trace_args.data());
// continue as normal with 'argv'
```
Unfortunately, I don't think we can just make `argv` a `const` since it must be operated on twice before calling `PROGRAM_START`.

Reply:
Sorry, I didn't realize that `shouldRunAsSubcommand` was already modifying the input arguments. I feel like I'd change that to just taking the input arguments as an immutable ArrayRef, and then making the copy inside the body of the `if`:

```
if (shouldRunAsSubcommand(ExecName, SubcommandName, argv)) {
  SmallVector<const char *, 256> subCommandArgs(argv.begin(), argv.end());
  subCommandArgs.erase(1);
  // …
```

What do you think for that?

(Sorry for the delayed responses. The run-up to WWDC has been slamming a bunch of us!)

Reply:
@jrose-apple About WWDC: no worries!

I attempted to implement your suggestion, and created a commit for it. It was necessary to factor out the lower half of `main` in order to properly handle the `repl` case. If the sub-command is `repl`, `shouldRunAsSubcommand` modifies `argv` and it returns `false`, so a modified version of `argv` is used outside of the `if` block.